File: src/main/java/com/music/music_inventory_api/service/AlbumService.java
package com.music.music_inventory_api.service;

import com.music.music_inventory_api.dto.request.CreateAlbumRequest;
import com.music.music_inventory_api.dto.request.UpdateAlbumRequest;
import com.music.music_inventory_api.dto.response.AlbumDetailResponse;
import com.music.music_inventory_api.dto.response.AlbumResponse;
import java.math.BigDecimal;
import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for Album business logic.
 */
public interface AlbumService
{

    /**
     * Create a new album.
     *
     * @param request
     *            the album creation request
     * @return the created album response
     */
    AlbumResponse createAlbum(CreateAlbumRequest request);

    /**
     * Get album by ID with detailed information including songs.
     *
     * @param id
     *            the album ID
     * @return the detailed album response
     */
    AlbumDetailResponse getAlbumById(Long id);

    /**
     * Get all albums with pagination.
     *
     * @param pageable
     *            pagination information
     * @return page of album responses
     */
    Page<AlbumResponse> getAllAlbums(Pageable pageable);

    /**
     * Search albums by title or artist name.
     *
     * @param searchTerm
     *            the search term
     * @return list of matching albums
     */
    List<AlbumResponse> searchAlbums(String searchTerm);

    /**
     * Get albums by genre.
     *
     * @param genreId
     *            the genre ID
     * @return list of albums in the genre
     */
    List<AlbumResponse> getAlbumsByGenre(Long genreId);

    /**
     * Get albums by price range.
     *
     * @param minPrice
     *            minimum price
     * @param maxPrice
     *            maximum price
     * @return list of albums in the price range
     */
    List<AlbumResponse> getAlbumsByPriceRange(BigDecimal minPrice, BigDecimal maxPrice);

    /**
     * Update an existing album.
     *
     * @param id
     *            the album ID
     * @param request
     *            the update request
     * @return the updated album response
     */
    AlbumResponse updateAlbum(Long id, UpdateAlbumRequest request);

    /**
     * Delete an album.
     *
     * @param id
     *            the album ID
     */
    void deleteAlbum(Long id);
}

File: frontend/src/services/albumService.ts
import api from './api';

export interface Artist {
    id: number;
    name: string;
    bio?: string;
    imageUrl?: string;
}

export interface Genre {
    id: number;
    name: string;
    description?: string;
}

export interface Album {
    id: number;
    title: string;
    artist: Artist;
    genres: Genre[];
    releaseDate: string;
    price: number;
    stockQuantity: number;
    coverImageUrl?: string;
    description?: string;
    format?: string;
    label?: string;
}

export interface CreateAlbumRequest {
    title: string;
    artistId: number;
    genreId: number;
    releaseDate: string;
    price: number;
    stockQuantity: number;
    imageUrl?: string;
}

export interface UpdateAlbumRequest {
    title?: string;
    artistId?: number;
    genreId?: number;
    releaseDate?: string;
    price?: number;
    stockQuantity?: number;
    imageUrl?: string;
}

class AlbumService {
    private readonly endpoint = '/albums';

    async getAll(): Promise<Album[]> {
        const response = await api.get<any>(this.endpoint);
        return response.data.content || response.data;
    }

    async getById(id: number): Promise<Album> {
        const response = await api.get<Album>(`${this.endpoint}/${id}`);
        return response.data;
    }

    async getByArtist(artistId: number): Promise<Album[]> {
        const response = await api.get<Album[]>(`${this.endpoint}/artist/${artistId}`);
        return response.data;
    }

    async getByGenre(genreId: number): Promise<Album[]> {
        const response = await api.get<Album[]>(`${this.endpoint}/genre/${genreId}`);
        return response.data;
    }

    async create(albumData: CreateAlbumRequest): Promise<Album> {
        const response = await api.post<Album>(this.endpoint, albumData);
        return response.data;
    }

    async update(id: number, albumData: UpdateAlbumRequest): Promise<Album> {
        const response = await api.put<Album>(`${this.endpoint}/${id}`, albumData);
        return response.data;
    }

    async delete(id: number): Promise<void> {
        await api.delete(`${this.endpoint}/${id}`);
    }
}

export default new AlbumService();
